### PR TITLE
[multistage] test restructure

### PR DESCRIPTION
**Pull request by walterddr**
step 1 of test restructure. 

- create mock routing manager factory (for planner)
- create mock data manager factory (for runtime)
- move the default set of test schemas, tables & segments into QueryEnvironmentTestBase (also used by QueryRunnerTestBase)
  - the creation of the simplest test table/data is completely clean from the mock routing/data manager and the query server enclosure. thus making them available for supporting other SQL and Data set (from test resources)

TODO
step 2 to split the test data providers into categories, and create their own specific test schemas and test cases.

### CODE REVIEW

File: pinot-query-planner/src/test/java/org/apache/pinot/query/testutils/QueryTestUtils.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import java.net.ServerSocket;
+
+
+public class QueryTestUtils {
+
+  private QueryTestUtils() {
+    // do not instantiate.
+  }
+
+  public static int getAvailablePort() {
```
Comment (agavra) [inline pinot-query-planner/src/test/java/org/apache/pinot/query/testutils/QueryTestUtils.java:31]: this should be `synchronized` (surprisingly, I've hit issues with this in a previous life)

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java [hunk 2]
```diff
@@ -157,6 +157,9 @@ private void compareRowEquals(List<Object[]> resultRows, List<Object[]> expected
   @DataProvider(name = "testDataWithSqlToFinalRowCount")
   private Object[][] provideTestSqlAndRowCount() {
     return new Object[][] {
+        // Hybrid table - hybrid table has a 1-day cut-off from current, so it will always use REALTIME only
```
Comment (agavra) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:160]: can we test a hybrid table that has data from both with this new refactor?

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTestBase.java [hunk 3]
```diff
@@ -127,29 +128,44 @@ protected void addDataToH2(Map<String, List<GenericRow>> rowsMap)
...
       throws Exception {
     DataTableBuilderFactory.setDataTableVersion(DataTableFactory.VERSION_4);
-    QueryServerEnclosure server1 = new QueryServerEnclosure(
-        ImmutableMap.of("a", INDEX_DIR_S1_A, "b", INDEX_DIR_S1_B, "c", INDEX_DIR_S1_C, "d_O", INDEX_DIR_S1_D),
-        QueryEnvironmentTestUtils.SERVER1_SEGMENTS);
-    QueryServerEnclosure server2 = new QueryServerEnclosure(
-        ImmutableMap.of("a", INDEX_DIR_S2_A, "c", INDEX_DIR_S2_C, "d_R", INDEX_DIR_S2_D, "d_O", INDEX_DIR_S1_D),
-        QueryEnvironmentTestUtils.SERVER2_SEGMENTS);
+    MockInstanceDataManagerFactory factory1 = new MockInstanceDataManagerFactory("server1")
+        .registerTable(QueryEnvironmentTestBase.SCHEMA_BUILDER.setSchemaName("a").build(), "a_REALTIME")
```
Comment (agavra) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTestBase.java:132]: tables must have segments, right? should we consider making them one API call? `registertable(schema, name, segment...)`

**Review by agavra — COMMENTED** [review]
nice! I like this change a lot, it makes building these kind of test cases much easier :)

File: pinot-query-planner/src/test/java/org/apache/pinot/query/QueryEnvironmentTestBase.java [hunk 4]
```diff
@@ -71,4 +98,42 @@ protected Object[][] provideQueries() {
             + " b.col2 NOT IN ('alice', 'charlie')"},
     };
   }
+
+  public static List<GenericRow> buildRows(String tableName) {
+    List<GenericRow> rows = new ArrayList<>(NUM_ROWS);
+    for (int i = 0; i < NUM_ROWS; i++) {
```
[the remark this answers is not in the document]
Comment (walterddr) [reply pinot-query-planner/src/test/java/org/apache/pinot/query/QueryEnvironmentTestBase.java:104]: good catch. will do

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTestBase.java (hunk 3 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTestBase.java:132]: table can have 0 segments

File: pinot-query-planner/src/test/java/org/apache/pinot/query/testutils/QueryTestUtils.java (hunk 1 above)
Comment (walterddr) [reply pinot-query-planner/src/test/java/org/apache/pinot/query/testutils/QueryTestUtils.java]: fixed this one, but also added a java doc explaining this is still not threadsafe/potential race condition can still occur

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/OperatorTestUtil.java [hunk 5]
```diff
@@ -18,33 +18,43 @@
...
-    return getEndOfStreamRowBlockWithSchema();
+  public static final String OP_1 = "op1";
+  public static final String OP_2 = "op2";
+
+  static {
+    MOCK_OPERATOR_FACTORY = new MockDataBlockOperatorFactory()
+        .registerOperator(OP_1, SIMPLE_KV_DATA_SCHEMA)
+        .registerOperator(OP_2, SIMPLE_KV_DATA_SCHEMA)
+        .addRows(OP_1, SIMPLE_KV_DATA_ROWS.get(0))
+        .addRows(OP_2, SIMPLE_KV_DATA_ROWS.get(1));
```
Comment (walterddr) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/OperatorTestUtil.java:47]: FYI @61yao this would be how I envision ppl use this data block operator to add test data.

also add multiple rows means return multiple data blocks (this would be super useful to test more complex cases)

Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/OperatorTestUtil.java:47]: also a sidenote. i think this static block should be in each test files (since they can register different data schema / rows)

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTestBase.java [hunk 6]
```diff
@@ -127,29 +128,44 @@ protected void addDataToH2(Map<String, List<GenericRow>> rowsMap)
   public void setUp()
       throws Exception {
     DataTableBuilderFactory.setDataTableVersion(DataTableFactory.VERSION_4);
-    QueryServerEnclosure server1 = new QueryServerEnclosure(
-        ImmutableMap.of("a", INDEX_DIR_S1_A, "b", INDEX_DIR_S1_B, "c", INDEX_DIR_S1_C, "d_O", INDEX_DIR_S1_D),
-        QueryEnvironmentTestUtils.SERVER1_SEGMENTS);
-    QueryServerEnclosure server2 = new QueryServerEnclosure(
-        ImmutableMap.of("a", INDEX_DIR_S2_A, "c", INDEX_DIR_S2_C, "d_R", INDEX_DIR_S2_D, "d_O", INDEX_DIR_S1_D),
-        QueryEnvironmentTestUtils.SERVER2_SEGMENTS);
+    MockInstanceDataManagerFactory factory1 = new MockInstanceDataManagerFactory("server1")
```
Comment (walterddr) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTestBase.java:131]: at the current moment this is stil manually registering using static code/member variable. I was thinking we should actually load data schema / test data (csv or other format) from file. thoughts?

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+public class MockDataBlockOperatorFactory {
+  protected final Map<String, List<List<Object[]>>> _rowsMap;
+  protected final Map<String, DataSchema> _operatorSchemaMap;
+
+  public MockDataBlockOperatorFactory() {
+    _rowsMap = new HashMap<>();
+    _operatorSchemaMap = new HashMap<>();
+  }
+
+  public MockDataBlockOperatorFactory registerOperator(String operatorName, DataSchema resultSchema) {
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java:44]: Don't we need to register the input schema as well for the operator ?

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    _operatorSchemaMap.put(operatorName, resultSchema);
+    return this;
+  }
+
+  public MockDataBlockOperatorFactory addRows(String operatorName, List<Object[]> rows) {
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java:49]: This are the rows input / consumed by the operator right ?

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockInstanceDataManagerFactory.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    if (tableType == null) {
+      registerTableNameWithType(schema, TableNameBuilder.forType(TableType.OFFLINE).tableNameWithType(tableName));
+      registerTableNameWithType(schema, TableNameBuilder.forType(TableType.REALTIME).tableNameWithType(tableName));
+    } else {
+      registerTableNameWithType(schema, tableName);
+    }
+    return this;
+  }
+
+  public MockInstanceDataManagerFactory addSegment(String tableNameWithType, List<GenericRow> rows) {
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockInstanceDataManagerFactory.java:83]: I am not sure why is this needed. Segment level execution will anyway be done by leaf layers (current engine operators). For that we already have a pretty good framework for doing segment level (operator level) or inter segment / inter server type of testing.

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  @SuppressWarnings("unchecked")
+  public BaseOperator<TransferableBlock> buildMockOperator(String operatorName) {
+    BaseOperator<TransferableBlock> operator = Mockito.mock(BaseOperator.class);
+    Mockito.when(operator.nextBlock()).thenAnswer(new Answer<Object>() {
+      private int _invocationCount = 0;
+      public Object answer(InvocationOnMock invocation) {
+        return _invocationCount >= _rowsMap.get(operatorName).size()
+            ? TransferableBlockUtils.getEndOfStreamTransferableBlock()
+            : new TransferableBlock(_rowsMap.get(operatorName).get(_invocationCount++),
+                _operatorSchemaMap.get(operatorName), BaseDataBlock.Type.ROW);
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java:65]: This seems like a pass-through implementation. Assuming the rowsMap is the input set of rows, shouldn't the mock operator need to process them as part of `nextBlock` ?

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java (hunk 7 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java:44]: yes. the 2nd argument is a schema. i don't foresee we need to separately register operator and schema.

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java (hunk 8 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java:49]: correct this is the mock data being returned by the mock operator

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java (hunk 10 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java:65]: this method build a mock operator that returns sequentially what's being added during factory construction.

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+  public MockDataBlockOperatorFactory addRows(String operatorName, List<Object[]> rows) {
+    List<List<Object[]>> operatorRows = _rowsMap.getOrDefault(operatorName, new ArrayList<>());
+    operatorRows.add(rows);
+    _rowsMap.put(operatorName, operatorRows);
+    return this;
+  }
+
+  @SuppressWarnings("unchecked")
+  public BaseOperator<TransferableBlock> buildMockOperator(String operatorName) {
```
Comment (61yao) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java:57]: Is this returning one row at a time if I read it right?

**Review by agavra — APPROVED** [review]
LGTM! some style nits that you can feel free to ignore. thanks @walterddr

File: pinot-query-planner/src/test/java/org/apache/pinot/query/testutils/MockRoutingManagerFactory.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    _tableServerSegmentMap.put(tableNameWithType, serverSegmentMap);
+    return this;
+  }
+
+  public RoutingManager buildRoutingManager() {
+    // create all the mock routing tables
+    _mockRoutingTableMap.clear();
+    for (Map.Entry<String, Map<ServerInstance, List<String>>> tableEntry : _tableServerSegmentMap.entrySet()) {
+      String tableNameWithType = tableEntry.getKey();
+      RoutingTable mockRoutingTable = mock(RoutingTable.class);
```
[the remark this answers is not in the document]
Comment (walterddr) [reply pinot-query-planner/src/test/java/org/apache/pinot/query/testutils/MockRoutingManagerFactory.java:102]: good call out. let me try to do that. the only thing is i don't know what's the other 2 fields are suppose to be but yeah for now I will fill blank

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java (hunk 11 above)
Comment (61yao) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java:57]: I see. Makes sense.

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/OperatorTestUtil.java (hunk 5 above)
Comment (61yao) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/OperatorTestUtil.java:47]: This is really useful!

**Comment by 61yao** [comment]
I feel I understand how to use MockDataBlockOperatorFactory since there is an example. I am not sure about the other part. Would be great to see follow up example. Thanks a lot for doing this. This is a great help for us!

**Review by siddharthteotia — APPROVED** [review]
I still have a few questions on the usage and further simplification but not worth blocking

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java (hunk 7 above)
Comment (siddharthteotia) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java:44]: I didn't mean to register operator and schema separately. 

What I meant was that do we need to register operator along with input schema as well. This API registers operator + resultSchema. I meant operator + inputSchema + resultSchema for the cases where operators produces output with slightly different schema than inputSchema.

Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/testutils/MockDataBlockOperatorFactory.java:44]: oh. i see the confusion. this class only returns a `mockOperator` with a mocked-up return when calling `nextBlock()` this is used in the test by feeding into the actual operator's constructor as the `upstreamOperator` 

so. the mock operator doesn't need to register input, as they are fake and stubbed out the return with a preset block anyway.

**Comment by walterddr** [comment]
> I feel I understand how to use MockDataBlockOperatorFactory since there is an example. I am not sure about the other part. Would be great to see follow up example. Thanks a lot for doing this. This is a great help for us!

good question. yeah b/c the next step is to split the current `QueryRunnerTest` and `QueryEnvironmentTest` into individual functionality test groups. there we shall see the usage, but should be very similar to the current usage (except register way less tables and data). 

but as of now this PR is already large enough so i decided to spill that into a separate PR. will create that one once this is merged.
